lib/jekyll/commands/serve/servlet.rb

```diff
@@ -21,10 +21,37 @@ def initialize(server, root, callbacks)
         # up with a different preference on which comes first.
 
         def search_file(req, res, basename)
-          # /file.* > /file/index.html > /file.html
+          # /file.* -> /file.html
           super || super(req, res, "#{basename}.html")
         end
 
+        def search_index_file(req, res)
+          # /file/index.html -> /file.html
+
+          # First, let's see if the our super method can figure it out.
+          # (i.e Check for index.html files in the res.filename directory)
+          file = super
+
+          unless file
+            # Ok, I guess that didn't work, I guess there's no basename/index.html
+            # Let's look for basename.html instead...
+
+            # Split the basename and split it from the path (res.filename)
+            # Regex matches a / that isn't followed by another / or EOL
+            index = res.filename.rindex %r!/(?!/|$)!
+            basename = res.filename[index..-1]
+            res.filename = res.filename[0, index]
+
+            # Try and find a file named dirname.html in the parent directory.
+            file = search_file(req, res, basename + ".html")
+
+            # If we didn't find a file, revert our changes to res.filename.
+            res.filename << basename unless file
+          end
+
+          return file
```

Comment:
you technically don't need the `return` statement, since ruby returns the last statement in a function anyways